Nodejs 12

Error when trying to install using
node: v12.1.0
npm: 6.9.0

```
$ npm install

> java@0.11.0 install /Users/shawn/tmp-thing/node_modules/java
> node-gyp rebuild

make: Entering directory '/Users/shawn/tmp-thing/node_modules/java/build'
  CXX(target) Release/obj.target/nodejavabridge_bindings/src/java.o
In file included from ../src/java.cpp:1:
In file included from ../src/java.h:9:
In file included from ../node_modules/nan/nan.h:221:
In file included from ../node_modules/nan/nan_converters.h:67:
../node_modules/nan/nan_converters_43_inl.h:22:1: warning: 'ToBoolean' is deprecated: ToBoolean can never throw. Use Local version. [-Wdeprecated-declarations]
X(Boolean)
^
../node_modules/nan/nan_converters_43_inl.h:18:12: note: expanded from macro 'X'
      val->To ## TYPE(isolate->GetCurrentContext())                            \
           ^
<scratch space>:175:1: note: expanded from here
ToBoolean
^
/Users/shawn/.node-gyp/12.1.0/include/node/v8.h:2523:3: note: 'ToBoolean' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("ToBoolean can never throw. Use Local version.",
  ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
In file included from ../src/java.cpp:1:
In file included from ../src/java.h:9:
In file included from ../node_modules/nan/nan.h:221:
In file included from ../node_modules/nan/nan_converters.h:67:
../node_modules/nan/nan_converters_43_inl.h:40:1: warning: 'BooleanValue' is deprecated: BooleanValue can never throw. Use Isolate version. [-Wdeprecated-declarations]
X(bool, Boolean)
^
../node_modules/nan/nan_converters_43_inl.h:37:15: note: expanded from macro 'X'
  return val->NAME ## Value(isolate->GetCurrentContext());                     \
              ^
<scratch space>:182:1: note: expanded from here
BooleanValue
^
/Users/shawn/.node-gyp/12.1.0/include/node/v8.h:2561:3: note: 'BooleanValue' has been explicitly marked deprecated here
  V8_DEPRECATED("BooleanValue can never throw. Use Isolate version.",
  ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8config.h:307:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
In file included from ../src/java.cpp:1:
In file included from ../src/java.h:9:
In file included from ../node_modules/nan/nan.h:222:
In file included from ../node_modules/nan/nan_new.h:189:
../node_modules/nan/nan_implementation_12_inl.h:103:42: error: no viable conversion from 'v8::Isolate *' to 'Local<v8::Context>'
  return scope.Escape(v8::Function::New( isolate
                                         ^~~~~~~
/Users/shawn/.node-gyp/12.1.0/include/node/v8.h:183:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'v8::Isolate *' to 'const v8::Local<v8::Context> &' for 1st argument
class Local {
      ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8.h:183:7: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'v8::Isolate *' to 'v8::Local<v8::Context> &&' for 1st argument
/Users/shawn/.node-gyp/12.1.0/include/node/v8.h:187:13: note: candidate template ignored: could not match 'Local<type-parameter-0-0>' against 'v8::Isolate *'
  V8_INLINE Local(Local<S> that)
            ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8.h:4126:22: note: passing argument to parameter 'context' here
      Local<Context> context, FunctionCallback callback,
                     ^
In file included from ../src/java.cpp:1:
In file included from ../src/java.h:9:
../node_modules/nan/nan.h:1064:44: warning: 'ToString' is deprecated: Use maybe version [-Wdeprecated-declarations]
      v8::Local<v8::String> string = from->ToString(v8::Isolate::GetCurrent());
                                           ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8.h:2543:3: note: 'ToString' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version",
  ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
In file included from ../src/java.cpp:1:
In file included from ../src/java.h:9:
../node_modules/nan/nan.h:1855:28: warning: 'Set' is deprecated: Use maybe version [-Wdeprecated-declarations]
    New(persistentHandle)->Set(New(key).ToLocalChecked(), value);
                           ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8.h:3358:3: note: 'Set' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version",
  ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
In file included from ../src/java.cpp:1:
In file included from ../src/java.h:9:
../node_modules/nan/nan.h:1861:28: warning: 'Set' is deprecated: Use maybe version [-Wdeprecated-declarations]
    New(persistentHandle)->Set(key, value);
                           ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8.h:3358:3: note: 'Set' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version",
  ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
In file included from ../src/java.cpp:1:
In file included from ../src/java.h:9:
../node_modules/nan/nan.h:1867:28: warning: 'Set' is deprecated: Use maybe version [-Wdeprecated-declarations]
    New(persistentHandle)->Set(index, value);
                           ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8.h:3367:3: note: 'Set' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version",
  ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
In file included from ../src/java.cpp:1:
In file included from ../src/java.h:9:
../node_modules/nan/nan.h:1873:32: warning: 'Get' is deprecated: Use maybe version [-Wdeprecated-declarations]
        New(persistentHandle)->Get(New(key).ToLocalChecked()));
                               ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8.h:3412:3: note: 'Get' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version", Local<Value> Get(Local<Value> key));
  ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
In file included from ../src/java.cpp:1:
In file included from ../src/java.h:9:
../node_modules/nan/nan.h:1879:48: warning: 'Get' is deprecated: Use maybe version [-Wdeprecated-declarations]
    return scope.Escape(New(persistentHandle)->Get(key));
                                               ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8.h:3412:3: note: 'Get' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version", Local<Value> Get(Local<Value> key));
  ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
In file included from ../src/java.cpp:1:
In file included from ../src/java.h:9:
../node_modules/nan/nan.h:1884:48: warning: 'Get' is deprecated: Use maybe version [-Wdeprecated-declarations]
    return scope.Escape(New(persistentHandle)->Get(index));
                                               ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8.h:3416:3: note: 'Get' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version", Local<Value> Get(uint32_t index));
  ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
In file included from ../src/java.cpp:1:
In file included from ../src/java.h:9:
In file included from ../node_modules/nan/nan.h:2698:
../node_modules/nan/nan_object_wrap.h:24:25: error: no member named 'IsNearDeath' in 'Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> >'
    assert(persistent().IsNearDeath());
           ~~~~~~~~~~~~ ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/assert.h:93:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)
                        ^
In file included from ../src/java.cpp:1:
In file included from ../src/java.h:9:
In file included from ../node_modules/nan/nan.h:2698:
../node_modules/nan/nan_object_wrap.h:67:18: warning: 'MarkIndependent' is deprecated: Weak objects are always considered independent. Use TracedGlobal when trying to use EmbedderHeapTracer. Use a strong handle when trying to keep an object alive.
      [-Wdeprecated-declarations]
    persistent().MarkIndependent();
                 ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8.h:563:3: note: 'MarkIndependent' has been explicitly marked deprecated here
  V8_DEPRECATED(
  ^
/Users/shawn/.node-gyp/12.1.0/include/node/v8config.h:307:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
In file included from ../src/java.cpp:1:
In file included from ../src/java.h:9:
In file included from ../node_modules/nan/nan.h:2698:
../node_modules/nan/nan_object_wrap.h:124:26: error: no member named 'IsNearDeath' in 'Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> >'
    assert(wrap->handle_.IsNearDeath());
           ~~~~~~~~~~~~~ ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/assert.h:93:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)
                        ^
In file included from ../src/java.cpp:1:
../src/java.h:19:24: error: no template named 'Handle' in namespace 'v8'
  static void Init(v8::Handle<v8::Object> target);
                   ~~~~^
In file included from ../src/java.cpp:7:
../src/javaObject.h:16:24: error: no template named 'Handle' in namespace 'v8'
  static void Init(v8::Handle<v8::Object> target);
                   ~~~~^
../src/java.cpp:52:23: error: variable has incomplete type 'void'
/*static*/ void Java::Init(v8::Handle<v8::Object> target) {
                      ^
../src/java.cpp:52:49: error: expected '(' for function-style cast or type construction
/*static*/ void Java::Init(v8::Handle<v8::Object> target) {
                                      ~~~~~~~~~~^
../src/java.cpp:52:32: error: no member named 'Handle' in namespace 'v8'
/*static*/ void Java::Init(v8::Handle<v8::Object> target) {
                           ~~~~^
../src/java.cpp:52:51: error: use of undeclared identifier 'target'
/*static*/ void Java::Init(v8::Handle<v8::Object> target) {
                                                  ^
../src/java.cpp:52:58: error: expected ';' after top level declarator
/*static*/ void Java::Init(v8::Handle<v8::Object> target) {
                                                         ^
                                                         ;
10 warnings and 10 errors generated.```

--- a/.travis.yml
+++ b/.travis.yml
@@ -8,6 +8,7 @@ env:
   - NODE_VERSION=7 CC=clang CXX=clang++
   - NODE_VERSION=8 CC=clang CXX=clang++
   - NODE_VERSION=11 CC=clang CXX=clang++
+  - NODE_VERSION=12 CC=clang CXX=clang++
 before_install:
   - nvm install $NODE_VERSION
 before_script:

--- a/package.json
+++ b/package.json
@@ -30,7 +30,7 @@
     "find-java-home": "0.2.0",
     "glob": "7.1.3",
     "lodash": "4.17.14",
-    "nan": "2.11.1"
+    "nan": "2.14.0"
   },
   "devDependencies": {
     "chalk": "2.4.1",

--- a/src/java.cpp
+++ b/src/java.cpp
@@ -49,7 +49,7 @@ bool v8ThreadIdEquals(threadId a, threadId b) {
 #endif
 }
 
-/*static*/ void Java::Init(v8::Handle<v8::Object> target) {
+/*static*/ void Java::Init(v8::Local<v8::Object> target) {
   Nan::HandleScope scope;
 
   v8ThreadId = my_getThreadId();
@@ -79,7 +79,7 @@ bool v8ThreadIdEquals(threadId a, threadId b) {
   Nan::SetPrototypeMethod(t, "setStaticFieldValue", setStaticFieldValue);
   Nan::SetPrototypeMethod(t, "instanceOf", instanceOf);
 
-  target->Set(Nan::New<v8::String>("Java").ToLocalChecked(), t->GetFunction());
+  Nan::Set(target, Nan::New<v8::String>("Java").ToLocalChecked(), Nan::GetFunction(t).ToLocalChecked());
 
   JavaProxyObject::init();
 }
@@ -90,10 +90,10 @@ NAN_METHOD(Java::New) {
   Java *self = new Java();
   self->Wrap(info.This());
 
-  self->handle()->Set(Nan::New<v8::String>("classpath").ToLocalChecked(), Nan::New<v8::Array>());
-  self->handle()->Set(Nan::New<v8::String>("options").ToLocalChecked(), Nan::New<v8::Array>());
-  self->handle()->Set(Nan::New<v8::String>("nativeBindingLocation").ToLocalChecked(), Nan::New<v8::String>("Not Set").ToLocalChecked());
-  self->handle()->Set(Nan::New<v8::String>("asyncOptions").ToLocalChecked(), Nan::Null());
+  Nan::Set(self->handle(), Nan::New<v8::String>("classpath").ToLocalChecked(), Nan::New<v8::Array>());
+  Nan::Set(self->handle(), Nan::New<v8::String>("options").ToLocalChecked(), Nan::New<v8::Array>());
+  Nan::Set(self->handle(), Nan::New<v8::String>("nativeBindingLocation").ToLocalChecked(), Nan::New<v8::String>("Not Set").ToLocalChecked());
+  Nan::Set(self->handle(), Nan::New<v8::String>("asyncOptions").ToLocalChecked(), Nan::Null());
 
   info.GetReturnValue().Set(info.This());
 }
@@ -133,29 +133,31 @@ void Java::configureAsync(v8::Local<v8::Value>& asyncOptions) {
   doAsync = false;
   doPromise = false;
 
-  v8::Local<v8::Value> suffixValue = asyncOptionsObj->Get(Nan::New<v8::String>("syncSuffix").ToLocalChecked());
-  if (suffixValue->IsString()) {
-    v8::Local<v8::String> suffix = suffixValue->ToString();
-    v8::String::Utf8Value utf8(suffix);
+  v8::MaybeLocal<v8::Value> maybeSuffixValue = Nan::Get(asyncOptionsObj, Nan::New<v8::String>("syncSuffix").ToLocalChecked());
+  v8::Local<v8::Value> suffixValue;
+  if (maybeSuffixValue.ToLocal(&suffixValue) && suffixValue->IsString()) {
+    v8::Local<v8::String> suffix = suffixValue->ToString(Nan::GetCurrentContext()).ToLocalChecked();
+    Nan::Utf8String utf8(suffix);
     m_SyncSuffix.assign(*utf8);
     doSync = true;
   }
 
-  suffixValue = asyncOptionsObj->Get(Nan::New<v8::String>("asyncSuffix").ToLocalChecked());
-  if (suffixValue->IsString()) {
-    v8::Local<v8::String> suffix = suffixValue->ToString();
-    v8::String::Utf8Value utf8(suffix);
+  maybeSuffixValue = Nan::Get(asyncOptionsObj, Nan::New<v8::String>("asyncSuffix").ToLocalChecked());
+  if (maybeSuffixValue.ToLocal(&suffixValue) && suffixValue->IsString()) {
+    v8::Local<v8::String> suffix = suffixValue->ToString(Nan::GetCurrentContext()).ToLocalChecked();
+    Nan::Utf8String utf8(suffix);
     m_AsyncSuffix.assign(*utf8);
     doAsync = true;
   }
 
-  suffixValue = asyncOptionsObj->Get(Nan::New<v8::String>("promiseSuffix").ToLocalChecked());
-  if (suffixValue->IsString()) {
-    v8::Local<v8::String> suffix = suffixValue->ToString();
-    v8::String::Utf8Value utf8(suffix);
+  maybeSuffixValue = Nan::Get(asyncOptionsObj, Nan::New<v8::String>("promiseSuffix").ToLocalChecked());
+  if (maybeSuffixValue.ToLocal(&suffixValue) && suffixValue->IsString()) {
+    v8::Local<v8::String> suffix = suffixValue->ToString(Nan::GetCurrentContext()).ToLocalChecked();
+    Nan::Utf8String utf8(suffix);
     m_PromiseSuffix.assign(*utf8);
-    v8::Local<v8::Value> promisify = asyncOptionsObj->Get(Nan::New<v8::String>("promisify").ToLocalChecked());
-    if (!promisify->IsFunction()) {
+    v8::MaybeLocal<v8::Value> maybePromisify = Nan::Get(asyncOptionsObj, Nan::New<v8::String>("promisify").ToLocalChecked());
+    v8::Local<v8::Value> promisify;
+    if (maybePromisify.ToLocal(&promisify) && !promisify->IsFunction()) {
       fprintf(stderr, "asyncOptions.promisify must be a function");
       assert(promisify->IsFunction());
     }
@@ -176,8 +178,9 @@ void Java::configureAsync(v8::Local<v8::Value>& asyncOptions) {
 }
 
 v8::Local<v8::Value> Java::createJVM(JavaVM** jvm, JNIEnv** env) {
-  v8::Local<v8::Value> asyncOptions = this->handle()->Get(Nan::New<v8::String>("asyncOptions").ToLocalChecked());
-  if (asyncOptions->IsObject()) {
+  v8::MaybeLocal<v8::Value> maybeAsyncOptions = Nan::Get(this->handle(), Nan::New<v8::String>("asyncOptions").ToLocalChecked());
+  v8::Local<v8::Value> asyncOptions;
+  if (maybeAsyncOptions.ToLocal(&asyncOptions) && asyncOptions->IsObject()) {
     configureAsync(asyncOptions);
   }
 
@@ -185,8 +188,9 @@ v8::Local<v8::Value> Java::createJVM(JavaVM** jvm, JNIEnv** env) {
   std::ostringstream classPath;
   classPath << "-Djava.class.path=";
 
-  v8::Local<v8::Value> classPathValue = this->handle()->Get(Nan::New<v8::String>("classpath").ToLocalChecked());
-  if(!classPathValue->IsArray()) {
+  v8::MaybeLocal<v8::Value> maybeClassPathValue = Nan::Get(this->handle(), Nan::New<v8::String>("classpath").ToLocalChecked());
+  v8::Local<v8::Value> classPathValue;
+  if(!maybeClassPathValue.ToLocal(&classPathValue) || !classPathValue->IsArray()) {
     return Nan::TypeError("Classpath must be an array");
   }
   v8::Local<v8::Array> classPathArrayTemp = v8::Local<v8::Array>::Cast(classPathValue);
@@ -203,18 +207,18 @@ v8::Local<v8::Value> Java::createJVM(JavaVM** jvm, JNIEnv** env) {
     if(!arrayItemValue->IsString()) {
       return Nan::TypeError("Classpath must only contain strings");
     }
-    v8::Local<v8::String> arrayItem = arrayItemValue->ToString();
-    v8::String::Utf8Value arrayItemStr(arrayItem);
+    v8::Local<v8::String> arrayItem = arrayItemValue->ToString(Nan::GetCurrentContext()).ToLocalChecked();
+    Nan::Utf8String arrayItemStr(arrayItem);
     classPath << *arrayItemStr;
   }
 
   // set the native binding location
-  v8::Local<v8::Value> v8NativeBindingLocation = this->handle()->Get(Nan::New<v8::String>("nativeBindingLocation").ToLocalChecked());
-  v8::String::Utf8Value nativeBindingLocationStr(v8NativeBindingLocation);
+  v8::Local<v8::Value> v8NativeBindingLocation = Nan::Get(this->handle(), Nan::New<v8::String>("nativeBindingLocation").ToLocalChecked()).FromMaybe(v8::Local<v8::Value>());
+  Nan::Utf8String nativeBindingLocationStr(v8NativeBindingLocation);
   s_nativeBindingLocation = *nativeBindingLocationStr;
 
   // get other options
-  v8::Local<v8::Value> optionsValue = this->handle()->Get(Nan::New<v8::String>("options").ToLocalChecked());
+  v8::Local<v8::Value> optionsValue = Nan::Get(this->handle(), Nan::New<v8::String>("options").ToLocalChecked()).FromMaybe(v8::Local<v8::Value>());
   if(!optionsValue->IsArray()) {
     return Nan::TypeError("options must be an array");
   }
@@ -232,8 +236,8 @@ v8::Local<v8::Value> Java::createJVM(JavaVM** jvm, JNIEnv** env) {
       delete[] vmOptions;
       return Nan::TypeError("options must only contain strings");
     }
-    v8::Local<v8::String> arrayItem = arrayItemValue->ToString();
-    v8::String::Utf8Value arrayItemStr(arrayItem);
+    v8::Local<v8::String> arrayItem = arrayItemValue->ToString(Nan::GetCurrentContext()).ToLocalChecked();
+    Nan::Utf8String arrayItemStr(arrayItem);
     vmOptions[i+1].optionString = strdup(*arrayItemStr);
   }
 
@@ -266,7 +270,7 @@ v8::Local<v8::Value> Java::createJVM(JavaVM** jvm, JNIEnv** env) {
 
   m_classLoader = getSystemClassLoader(*env);
 
-  v8::Local<v8::Value> onJvmCreated = this->handle()->Get(Nan::New<v8::String>("onJvmCreated").ToLocalChecked());
+  v8::Local<v8::Value> onJvmCreated = Nan::Get(this->handle(), Nan::New<v8::String>("onJvmCreated").ToLocalChecked()).FromMaybe(v8::Local<v8::Value>());
 
   // TODO: this handles sets put doesn't prevent modifing the underlying data. So java.classpath.push will still work which is invalid.
   Nan::SetAccessor(this->handle(), Nan::New<v8::String>("classpath").ToLocalChecked(), AccessorProhibitsOverwritingGetter, AccessorProhibitsOverwritingSetter);
@@ -278,7 +282,7 @@ v8::Local<v8::Value> Java::createJVM(JavaVM** jvm, JNIEnv** env) {
   if (onJvmCreated->IsFunction()) {
     v8::Local<v8::Function> onJvmCreatedFunc = onJvmCreated.As<v8::Function>();
     v8::Local<v8::Object> context = Nan::New<v8::Object>();
-    onJvmCreatedFunc->Call(context, 0, NULL);
+    Nan::Call(onJvmCreatedFunc, context, 0, NULL);
   }
 
   return Nan::Null();
@@ -287,7 +291,7 @@ v8::Local<v8::Value> Java::createJVM(JavaVM** jvm, JNIEnv** env) {
 NAN_GETTER(Java::AccessorProhibitsOverwritingGetter) {
   Java* self = Nan::ObjectWrap::Unwrap<Java>(info.This());
   Nan::HandleScope scope;
-  v8::String::Utf8Value nameStr(property);
+  Nan::Utf8String nameStr(property);
   if(!strcmp("classpath", *nameStr)) {
     info.GetReturnValue().Set(Nan::New(self->m_classPathArray));
     return;
@@ -310,7 +314,7 @@ NAN_GETTER(Java::AccessorProhibitsOverwritingGetter) {
 }
 
 NAN_SETTER(Java::AccessorProhibitsOverwritingSetter) {
-  v8::String::Utf8Value nameStr(property);
+  Nan::Utf8String nameStr(property);
   std::ostringstream errStr;
   errStr << "Cannot set " << *nameStr << " after calling any other java function.";
   Nan::ThrowError(errStr.str().c_str());
@@ -878,7 +882,7 @@ NAN_METHOD(Java::newArray) {
       env->SetObjectArrayElement((jobjectArray)results, i, val);
       if(env->ExceptionOccurred()) {
         std::ostringstream errStr;
-        v8::String::Utf8Value valStr(item);
+        Nan::Utf8String valStr(item);
         errStr << "Could not add item \"" << *valStr << "\" to array.";
         return Nan::ThrowError(javaExceptionToV8(self, env, errStr.str()));
       }
@@ -996,11 +1000,11 @@ NAN_METHOD(Java::newChar) {
   if(info[0]->IsNumber()) {
     charVal = (jchar)Nan::To<int32_t>(info[0]).FromJust();
   } else if(info[0]->IsString()) {
-    v8::Local<v8::String> val = info[0]->ToString();
+    v8::Local<v8::String> val = info[0]->ToString(Nan::GetCurrentContext()).ToLocalChecked();
     if(val->Length() != 1) {
       return Nan::ThrowError(Nan::TypeError("Argument 1 must be a string of 1 character."));
     }
-    std::string strVal = std::string(*v8::String::Utf8Value(val));
+    std::string strVal = std::string(*Nan::Utf8String(val));
     charVal = (jchar)strVal[0];
   } else {
     return Nan::ThrowError(Nan::TypeError("Argument 1 must be a number or string"));
@@ -1239,8 +1243,8 @@ void EIO_AfterCallJs(uv_work_t* req) {
 
   Nan::HandleScope scope;
   v8::Array* v8Args;
-  v8::Function* fn;
-  v8::Handle<v8::Value>* argv;
+  v8::Local<v8::Function> fn;
+  v8::Local<v8::Value>* argv;
   int argc;
   int i;
   v8::Local<v8::Value> v8Result;
@@ -1261,7 +1265,7 @@ void EIO_AfterCallJs(uv_work_t* req) {
     return;
   }
 
-  fn = v8::Function::Cast(*fnObj);
+  fn = fnObj.As<v8::Function>();
 
   if(dynamicProxyData->args) {
     v8Args = v8::Array::Cast(*javaArrayToV8(dynamicProxyData->java, env, dynamicProxyData->args));
@@ -1269,18 +1273,18 @@ void EIO_AfterCallJs(uv_work_t* req) {
   } else {
     argc = 0;
   }
-  argv = new v8::Handle<v8::Value>[argc];
+  argv = new v8::Local<v8::Value>[argc];
   for(i=0; i<argc; i++) {
     argv[i] = v8Args->Get(i);
   }
 
   Nan::TryCatch tryCatch;
   tryCatch.SetCaptureMessage(true);
-  v8Result = fn->Call(dynamicProxyDataFunctions, argc, argv);
+  v8Result = Nan::Call(fn, dynamicProxyDataFunctions, argc, argv).FromMaybe(v8::Local<v8::Value>());
   delete[] argv;
   if (tryCatch.HasCaught()) {
     dynamicProxyData->throwableClass = "node/NodeJsException";
-    v8::String::Utf8Value message(tryCatch.Message()->Get());
+    Nan::Utf8String message(tryCatch.Message()->Get());
     dynamicProxyData->throwableMessage = std::string(*message);
     tryCatch.Reset();
     dynamicProxyData->done = DYNAMIC_PROXY_JS_ERROR;

--- a/src/java.h
+++ b/src/java.h
@@ -16,7 +16,7 @@
 
 class Java : public Nan::ObjectWrap {
 public:
-  static void Init(v8::Handle<v8::Object> target);
+  static void Init(v8::Local<v8::Object> target);
   JavaVM* getJvm() { return m_jvm; }
   JNIEnv* getJavaEnv() { return m_env; } // can only be used safely by the main thread as this is the thread it belongs to
   jobject getClassLoader() { return m_classLoader; }

--- a/src/javaObject.cpp
+++ b/src/javaObject.cpp
@@ -7,7 +7,7 @@
 
 /*static*/ std::map<std::string, Nan::Persistent<v8::FunctionTemplate>*> JavaObject::sFunctionTemplates;
 
-/*static*/ void JavaObject::Init(v8::Handle<v8::Object> target) {
+/*static*/ void JavaObject::Init(v8::Local<v8::Object> target) {
 }
 
 /*static*/ v8::Local<v8::Object> JavaObject::New(Java *java, jobject obj) {
@@ -69,8 +69,8 @@
 
       if (java->DoPromise()) {
         v8::Local<v8::Object> recv = Nan::New<v8::Object>();
-        v8::Local<v8::Value> argv[] = { methodCallTemplate->GetFunction() };
-        v8::Local<v8::Value> result = promisify->Call(recv, 1, argv);
+        v8::Local<v8::Value> argv[] = { methodCallTemplate->GetFunction(Nan::GetCurrentContext()).ToLocalChecked() };
+        v8::Local<v8::Value> result = Nan::Call(promisify, recv, 1, argv).FromMaybe(v8::Local<v8::Value>());
         if (!result->IsFunction()) {
           fprintf(stderr, "Promisified result is not a function -- asyncOptions.promisify must return a function.\n");
           assert(result->IsFunction());
@@ -140,7 +140,7 @@ NAN_METHOD(JavaObject::methodCall) {
   JNIEnv *env = self->m_java->getJavaEnv();
   JavaScope javaScope(env);
 
-  v8::String::Utf8Value methodName(info.Data());
+  Nan::Utf8String methodName(info.Data());
   std::string methodNameStr = *methodName;
 
   int argsStart = 0;
@@ -176,7 +176,7 @@ NAN_METHOD(JavaObject::methodCallSync) {
   JNIEnv *env = self->m_java->getJavaEnv();
   JavaScope javaScope(env);
 
-  v8::String::Utf8Value methodName(info.Data());
+  Nan::Utf8String methodName(info.Data());
   std::string methodNameStr = *methodName;
 
   int argsStart = 0;
@@ -209,16 +209,18 @@ NAN_METHOD(JavaObject::methodCallSync) {
 NAN_METHOD(JavaObject::methodCallPromise) {
   Nan::HandleScope scope;
   v8::Local<v8::Function> fn = info.Data().As<v8::Function>();
-  v8::Handle<v8::Value>* argv = new v8::Handle<v8::Value>[info.Length()];
+  v8::Local<v8::Value>* argv = new v8::Local<v8::Value>[info.Length()];
   for (int i = 0 ; i < info.Length(); i++) {
     argv[i] = info[i];
   }
   
-  v8::Local<v8::Value> result = fn->Call(info.This(), info.Length(), argv);
+  v8::MaybeLocal<v8::Value> result = Nan::Call(fn, info.This(), info.Length(), argv);
   
   delete[] argv;
   
-  info.GetReturnValue().Set(result);
+  if (!result.IsEmpty()) {
+    info.GetReturnValue().Set(result.ToLocalChecked());
+  }
 }
 
 NAN_GETTER(JavaObject::fieldGetter) {
@@ -227,7 +229,7 @@ NAN_GETTER(JavaObject::fieldGetter) {
   JNIEnv *env = self->m_java->getJavaEnv();
   JavaScope javaScope(env);
 
-  v8::String::Utf8Value propertyCStr(property);
+  Nan::Utf8String propertyCStr(property);
   std::string propertyStr = *propertyCStr;
   jobject field = javaFindField(env, self->m_class, propertyStr);
   if(field == NULL) {
@@ -278,7 +280,7 @@ NAN_SETTER(JavaObject::fieldSetter) {
 
   jobject newValue = v8ToJava(env, value);
 
-  v8::String::Utf8Value propertyCStr(property);
+  Nan::Utf8String propertyCStr(property);
   std::string propertyStr = *propertyCStr;
   jobject field = javaFindField(env, self->m_class, propertyStr);
   if(field == NULL) {
@@ -345,7 +347,7 @@ NAN_INDEX_GETTER(JavaObject::indexGetter) {
 v8::Local<v8::Object> JavaProxyObject::New(Java *java, jobject obj, DynamicProxyData* dynamicProxyData) {
   Nan::EscapableHandleScope scope;
 
-  v8::Local<v8::Function> ctor = Nan::New(s_proxyCt)->GetFunction();
+  v8::Local<v8::Function> ctor = Nan::New(s_proxyCt)->GetFunction(Nan::GetCurrentContext()).ToLocalChecked();
   v8::Local<v8::Object> javaObjectObj = Nan::NewInstance(ctor).ToLocalChecked();
   SetHiddenValue(javaObjectObj, Nan::New<v8::String>(V8_HIDDEN_MARKER_JAVA_OBJECT).ToLocalChecked(), Nan::New<v8::Boolean>(true));
   JavaProxyObject *self = new JavaProxyObject(java, obj, dynamicProxyData);

--- a/src/javaObject.h
+++ b/src/javaObject.h
@@ -13,7 +13,7 @@ class Java;
 
 class JavaObject : public Nan::ObjectWrap {
 public:
-  static void Init(v8::Handle<v8::Object> target);
+  static void Init(v8::Local<v8::Object> target);
   static v8::Local<v8::Object> New(Java* java, jobject obj);
   static v8::Local<v8::Object> NewProxy(Java* java, jobject obj, DynamicProxyData* dynamicProxyData);
 

--- a/src/nodeJavaBridge.cpp
+++ b/src/nodeJavaBridge.cpp
@@ -3,7 +3,7 @@
 #include "javaObject.h"
 
 extern "C" {
-  static void init(v8::Handle<v8::Object> target) {
+  static void init(v8::Local<v8::Object> target) {
     Java::Init(target);
     JavaObject::Init(target);
   }

--- a/src/utils.cpp
+++ b/src/utils.cpp
@@ -339,7 +339,11 @@ jobject v8ToJava(JNIEnv* env, v8::Local<v8::Value> arg) {
   }
 
   if(arg->IsString()) {
+#if NODE_MAJOR_VERSION > 7
     v8::String::Value val(v8::Isolate::GetCurrent(), arg->ToString(Nan::GetCurrentContext()).ToLocalChecked());
+#else
+    v8::String::Value val(arg->ToString(Nan::GetCurrentContext()).ToLocalChecked());
+#endif
     return env->NewString(*val, val.length());
   }
 

--- a/src/utils.cpp
+++ b/src/utils.cpp
@@ -339,7 +339,7 @@ jobject v8ToJava(JNIEnv* env, v8::Local<v8::Value> arg) {
   }
 
   if(arg->IsString()) {
-    v8::String::Value val(arg->ToString());
+    v8::String::Value val(v8::Isolate::GetCurrent(), arg->ToString(Nan::GetCurrentContext()).ToLocalChecked());
     return env->NewString(*val, val.length());
   }
 
@@ -644,7 +644,7 @@ v8::Local<v8::Value> javaArrayToV8(Java* java, JNIEnv* env, jobjectArray objArra
   default:
     for(jsize i=0; i<arraySize; i++) {
         jobject obj = env->GetObjectArrayElement(objArray, i);
-        v8::Handle<v8::Value> item = javaToV8(java, env, obj);
+        v8::Local<v8::Value> item = javaToV8(java, env, obj);
         result->Set(i, item);
     }
     break;

--- a/src/utils.h
+++ b/src/utils.h
@@ -117,7 +117,7 @@ void unref(DynamicProxyData* dynamicProxyData);
     return;                                                                                  \
   }                                                                                          \
   v8::Local<v8::String> _##ARGNAME##_obj = v8::Local<v8::String>::Cast(info[argsStart]);     \
-  v8::String::Utf8Value _##ARGNAME##_val(_##ARGNAME##_obj);                                  \
+  Nan::Utf8String _##ARGNAME##_val(_##ARGNAME##_obj);                                  \
   std::string ARGNAME = *_##ARGNAME##_val;                                                   \
   argsStart++;
 
@@ -138,11 +138,11 @@ void unref(DynamicProxyData* dynamicProxyData);
 #define EXCEPTION_CALL_CALLBACK(JAVA, STRBUILDER) \
   std::ostringstream errStr;                                                            \
   errStr << STRBUILDER;                                                                 \
-  v8::Handle<v8::Value> error = javaExceptionToV8(JAVA, env, errStr.str());             \
-  v8::Handle<v8::Value> argv[2];                                                        \
+  v8::Local<v8::Value> error = javaExceptionToV8(JAVA, env, errStr.str());             \
+  v8::Local<v8::Value> argv[2];                                                        \
   argv[0] = error;                                                                      \
   argv[1] = Nan::Undefined();                                                           \
-  v8::Function::Cast(*callback)->Call(Nan::GetCurrentContext()->Global(), 2, argv);
+  Nan::Call(callback.As<v8::Function>(), Nan::GetCurrentContext()->Global(), 2, argv);
 
 #define END_CALLBACK_FUNCTION(MSG) \
   if(callbackProvided) {                                     \

